catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/security/accesscontrol/TagAggregationVisitor.java

```diff
@@ -0,0 +1,242 @@
+public class TagAggregationVisitor extends TagBaseVisitor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(TagAggregationVisitor.class);
+  private final Deque<TagAggregationRule> ruleDeque = new ArrayDeque<>();
+  private final Set<String> tags = new HashSet<>();
+
+  @Override
+  protected Object preProcessNonTagPredicate(Filter filter, Object data) {
+    reportNonTagEncountered();
+    return data;
+  }
+
+  public Set<String> getTags() {
+    return tags;
+  }
+
+  @Override
+  public Object visit(Not filter, Object data) {
+    ruleDeque.addFirst(TagAggregationRule.newRule());
+    data = super.visit(filter, data);
+    ruleDeque.removeFirst();
+    return data;
+  }
+
+  @Override
+  public Object visit(And filter, Object data) {
+    ruleDeque.addFirst(TagAggregationRule.newRule());
+    data = super.visit(filter, data);
+
+    TagAggregationRule rule = ruleDeque.removeFirst();
+    if (ruleDeque.isEmpty()) {
+      tags.addAll(rule.getTags());
+    } else {
+      ruleDeque.peekFirst().foundTags(rule.getTags());
+    }
+
+    return data;
+  }
+
+  @Override
+  public Object visit(Or filter, Object data) {
+    ruleDeque.addFirst(TagAggregationRule.newRule());
+    data = super.visit(filter, data);
+
+    TagAggregationRule rule = ruleDeque.removeFirst();
+    if (rule.isNotLimitedToTagCriteria()) {
+      // Cannot guarantee result set is a proper subset of policy
+      return data;
+    }
+
+    if (ruleDeque.isEmpty()) {
+      tags.addAll(rule.getTags());
+    } else {
+      ruleDeque.peekFirst().foundTags(rule.getTags());
+    }
+
+    return data;
+  }
+
+  @Override
+  public Object visit(PropertyIsLike filter, Object data) {
+    Map.Entry<String, String> pair =
+        filterAsKeyValuePair(filter.getExpression(), filter.getLiteral());
+    if (pair != null) {
+      LOGGER.trace("Found key-value predicate {}, looking for metacard tags", pair);
+      saveTag(pair.getKey(), pair.getValue());
+      return super.visit(filter, data);
+    }
+    LOGGER.trace(
+        "Filter {} could not be converted to key-value pair so could not be referring to metacard tags",
+        filter);
+    reportNonTagEncountered();
+    return super.visit(filter, data);
+  }
+
+  @Override
+  public Object visit(PropertyIsEqualTo filter, Object data) {
+    Map.Entry<String, String> pair =
+        filterAsKeyValuePair(filter.getExpression1(), filter.getExpression2());
+    if (pair != null) {
+      LOGGER.trace("Found key-value predicate {}, looking for metacard tags", pair);
+      saveTag(pair.getKey(), pair.getValue());
+      return super.visit(filter, data);
+    }
+    LOGGER.trace(
+        "Filter {} could not be converted to key-value pair so could not be referring to metacard tags",
+        filter);
+    reportNonTagEncountered();
+    return super.visit(filter, data);
+  }
+
+  private void saveTag(String property, String value) {
+    if (!Core.METACARD_TAGS.equals(property)) {
```

Comment:
So if the query has other criteria besides METACARD TAGS we cant apply security filter?

Reply:
It all depends what that criteria's relationship is with the tags criteria. If it's an `AND` relationship, sure. If it's an `OR` relationship, no we can't, because the result set of that query or subquery is not guaranteed to have the tag we're looking for. Same goes for `NOT` operators, since negative queries define their results based on what's not in them vs what's actually in them. So we can't make the assumption there either.